celery_unique/core.py

```diff
@@ -0,0 +1,199 @@
+        )
+
+    @staticmethod
+    def _make_ttl_for_unique_task_record(task_options):
+        """
+        Calculate an aproximate TTL for an enqueued task.
+
+        Given the options provided to `apply_async()` as keyword arguments,
+        determines the appropriate TTL to ensure that a unique task record
+        expires (approximately) at the same time as the earliest time that the
+        task is expected to be executed by Celery.
+
+        The TTL value will be determined by examining the following values, in
+        order of preference:
+            - The `eta` keyword argument passed to `apply_async()`, if any.  If
+              this value is found, then the TTL will be the number of seconds
+              between now and the ETA datetime.
+            - The `countdown` keyword argument passed to `apply_async()`, which
+              will theoretically always exist if `eta` was not provided.  If
+              this value is used, the TTL will be equal.
+
+        Additionally, if an `expires` keyword argument was passed, and its
+        value represents (either as an integer or timedelta) a shorter duration
+        of time than the values provided by `eta` or `countdown`, the TTL will
+        be reduced to the value of `countdown`.
+
+        Finally, the TTL value returned by this method will always be greater
+        than or equal to 1, in order to ensure compatibility with different
+        backend's TTL requirements, and that a record produced for a
+        nonexistent task will only live for a maximum of 1 second.
+
+        :param dict task_options: The values passed as additional keyword
+            arguments to `apply_async()`
+
+        :return: The TTL (in seconds) for the Redis record to-be-created
+        :rtype: int
+        """
+        # Set a default TTL as 1 second (in case actual TTL already occurred)
+        ttl_seconds = 1
+
+        option_keys = task_options.keys()
+        if 'eta' in option_keys:
+            # Get the difference between the ETA and now (relative to the ETA's timezone)
+            ttl_seconds = int(
+                (task_options['eta'] - datetime.datetime.now(tz=task_options['eta'].tzinfo)).total_seconds()
+            )
+        elif 'countdown' in option_keys:
+            ttl_seconds = task_options['countdown']
+
+        if 'expires' in option_keys:
+            if isinstance(task_options['expires'], datetime.datetime):
+                # Get the difference between the countdown and now (relative to the countdown's timezone)
+                seconds_until_expiry = int(
+                    (task_options['expires'] - datetime.datetime.now(tz=task_options['expires'].tzinfo)).total_seconds()
+                )
+            else:
+                seconds_until_expiry = task_options['expires']
+            if seconds_until_expiry < ttl_seconds:
+                ttl_seconds = seconds_until_expiry
+
+        if ttl_seconds <= 0:
```

Comment:
It shouldn't really ever happen that `ttl_seconds` is a value greater than 0 but less than 1, but it appears at least possible when using `task_options['countdown']`.

Instead this should be:
```python
if ttl_seconds < 1:
    ttl_seconds = 1
```

Probably more semantically correct anyways, and it ensures that our docstring's assertion that "the TTL value returned by this method will always be greater than or equal to 1" is actually true.